src/org/pentaho/reporting/platform/plugin/output/CachingPageableHTMLOutput.java

```diff
@@ -97,10 +97,21 @@ public CacheListener( final String key, final int acceptedPage,
     }
 
     @Override public void reportProcessingUpdate( final ReportProgressEvent reportProgressEvent ) {
-      //When we request X page the accepted-page is X-1
-      //then to be sure that X page is already stored the current page in event should be X+1
-      //in other words acceptedPage + 2
-      final boolean requestedPageIsStored = reportProgressEvent.getPage() == acceptedPage + 2;
+      //When we request X page the index that comes to server is X-1
+      //then to be sure that X page is already stored the current index in event should be X+1
+      //in other words page + 2
+      final int page;
+
+      if ( asyncReportListener.getRequestedPage() > 0 ) {
+        //user requested new page before original generation finished
+        //Generation could already pass the page but we need to update cache anyway
+        page = Math.max( asyncReportListener.getRequestedPage() + 2, reportProgressEvent.getPage() );
+      } else {
+        //it's just an original request
+        page = acceptedPage + 2;
+      }
+
+      final boolean requestedPageIsStored = reportProgressEvent.getPage() == page;
```

Comment:
You can get multiple progress events per page. It might be wise to have a flag that remembers the last page you persisted. This way you only persist a given page-set once. 

Adding a new member variable here for "lastAcceptedPageWritten" and checking against that as well as your accepted-page will save you from storing the same content multiple times.